DynamicTablesPkg/Library/Acpi/Common/AcpiMsctLib/MsctGenerator.c

```diff
@@ -0,0 +1,823 @@
+                     CfgMgrProtocol,
+                     &MaxClockDomain
+                     );
+  if (EFI_ERROR (Status) && (Status != EFI_NOT_FOUND)) {
+    goto error_handler;
+  }
+
+  Status = GetMemoryDomainInfo (
+             CfgMgrProtocol,
+             &MemDomainInfo,
+             &MemDomainInfoCount
+             );
+  if (EFI_ERROR (Status) && (Status != EFI_NOT_FOUND)) {
+    goto error_handler;
+  }
+
+  Status = GetProcessorDomainInfo (
+             CfgMgrProtocol,
+             &ProcDomainInfo,
+             &ProcDomainInfoCount
+             );
+  if (EFI_ERROR (Status) && (Status != EFI_NOT_FOUND)) {
+    goto error_handler;
+  }
+
+  DomainInfoCount = MemDomainInfoCount + ProcDomainInfoCount;
+  if (DomainInfoCount == 0) {
+    DEBUG ((
+      DEBUG_INFO,
+      "MSCT: No Memory/Processor Domain information found.\n"
+      ));
+    goto generate_msct_table;
+  }
+
+  DomainInfo = AllocateZeroPool (
+                 sizeof (EFI_ACPI_6_5_MAXIMUM_PROXIMITY_DOMAIN_INFORMATION_STRUCTURE) *
+                 DomainInfoCount
+                 );
+  if (DomainInfo == NULL) {
+    Status = EFI_OUT_OF_RESOURCES;
+    goto error_handler;
+  }
+
+  /// Get the maximum memory capacity per proximity domain.
+  for (MemIndex = 0; MemIndex <= MemDomainInfoCount; MemIndex++) {
+    if (MemDomainInfo[MemIndex].ProximityDomainRangeLow == MAX_UINT32) {
+      continue;
+    }
+
+    MaxProximityDomain = MAX (
+                           MaxProximityDomain,
+                           MemDomainInfo[MemIndex].ProximityDomainRangeHigh
+                           );
+
+    if (MemDomainInfo[MemIndex].ProximityDomainRangeLow ==
+        MemDomainInfo[MemIndex + 1].ProximityDomainRangeLow)
+    {
+      /// Maximum memory capacity per domain.
+      MemDomainInfo[MemIndex].MaximumMemoryCapacity +=
+        MemDomainInfo[MemIndex + 1].MaximumMemoryCapacity;
+      /// Invalidate the record
+      MemDomainInfo[MemIndex + 1].ProximityDomainRangeLow  = MAX_UINT32;
+      MemDomainInfo[MemIndex + 1].ProximityDomainRangeHigh = MAX_UINT32;
+    }
+  }
+
+  /// Perform union of MemDomainInfo and ProcDomainInfo for same proximity domain
+  DomainInfoIndex    = 0;
+  MaxProximityDomain = 0;
+  for (ProcIndex = 0; ProcIndex < ProcDomainInfoCount; ProcIndex++) {
+    if (ProcDomainInfo[ProcIndex].ProximityDomainRangeLow == MAX_UINT32) {
+      continue;
+    }
+
+    MaxProximityDomain = MAX (
+                           MaxProximityDomain,
+                           ProcDomainInfo[ProcIndex].ProximityDomainRangeHigh
+                           );
+
+    for (MemIndex = 0; MemIndex < MemDomainInfoCount; MemIndex++) {
+      if (MemDomainInfo[MemIndex].ProximityDomainRangeLow == MAX_UINT32) {
+        continue;
+      }
+
+      if (ProcDomainInfo[ProcIndex].ProximityDomainRangeLow ==
+          MemDomainInfo[MemIndex].ProximityDomainRangeLow)
+      {
+        CopyMem (
+          &DomainInfo[DomainInfoIndex],

```

Comment:
Isn't it possible to group the processor and memory information into one of the existing array, i.e. either `ProcDomainInfo` or `MemDomainInfo` ? That way we don't need to allocate the extra `DomainInfo` array.